babel.config.js

```diff
@@ -4,6 +4,8 @@ module.exports = {
       "@babel/preset-env",
       {
         targets: { node: 14 },
```

Comment:
this should be updated now :p

Reply:
ah, maybe it shouldnt - since we are only bumping the node version for the CLI itself

Reply:
I actually don't know `@preconstruct/cli`'s exact behavior here - does the build downlevelling use babel or TSConfig to decide what to output?

Reply:
Runtime files are exclusively~ transformed by Babel

> it seems like preconstruct build does not consider babel.config.js

It sounds like a configuration issue or a bug in Preconstruct. How did you come up with that conclusion? What I could try to repro locally and how I could observe the broken output?

Reply:
- In `@manypkg/cli` upgrade `package-json@^10.0.1`.
 - `yarn install`
 - `yarn preconstruct dev`
 - `yarn manypkg check` # broken
 - `yarn preconstruct build`
 - `yarn manypkg check` # works

Reply:
`yarn preconstruct dev` starts working only after you explicitly add `exclude: ["proposal-dynamic-import"],` to `babel.config.js`, while `yarn preconstruct build` works only if you have
```    
    "___experimentalFlags_WILL_CHANGE_IN_PATCH": {
      "keepDynamicImportAsDynamicImportInCommonJS": true
    }
```

Reply:
Relevant preconstruct code is that the experimental flag adds a rollup plugin:

https://github.com/preconstruct/preconstruct/blob/54c412d6f10752e3e4d0aa04aced1c44958baee1/packages/cli/src/build/rollup.ts#L221C1-L222C79

So, babel seems to be completely ignored in that regard, and I don't see anything that would get the preset-env babel target in that file either.

Reply:
It seems like the rollup setting that would control build output is `outputs.generatedCode`. It doesn't seem like `preconstruct` sets this explicitly, which probably means that the build uses the rollup defaults: 

https://github.com/rollup/rollup/blob/v2.79.2/src/utils/options/normalizeOutputOptions.ts#L346-L363

Reply:
nice github digging, my style - i love it :) The second link might be slightly inaccurate as it points to `master` an not to the tagged version but I doubt it matters here much (pro tip - you can hit `y` to get a permalink in the URL bar)

> It seems like the rollup setting that would control build output is outputs.generatedCode. It doesn't seem like preconstruct sets this explicitly, which probably means that the build uses the rollup defaults:

I'm slightly confused here given the plugin in Preconstruct that you linked to. But tbh this might just be the issue of not checking out locally how it behaves and what's the actual current output. Given you already looked into it - do you think that anything has to be fixed in Preconstruct itself? 

I'd treat the configuration issue with Babel for `preconstruct dev` purposes separately. This might not be something to fix in Preconstruct. So I'm primarily interested in the effect of all of this on `preconstruct build`

Reply:
Yeah, the second link is to master, [but it's the same for version 2.79.2](https://github.com/rollup/rollup/blob/v2.79.2/src/utils/options/options.ts#L86-L92).

As for preconstruct... I don't think preconstruct is doing anything *wrong*, per se. It's not obvious that you need to configure 2 separate settings for `build` and `dev` to execute in the same way, but it does make some sense. (tangentially, `keepDynamicImportAsDynamicImportInCommonJS` would be easier to implement in `rollup@3`, because it's now [a separate setting that defaults to `true`](https://rollupjs.org/migration/#dynamic-import-in-commonjs-output), so `preconstruct` wouldn't need a custom plugin.)

On the more broad question of what the output of `preconstruct build` should be - there doesn't seem to be a way to configure `preconstruct` to output more modern JS. `ES2015` is fully supported on `node >= 10.15.0` (and `node 8` is missing only like 2 methods), but the rollup `es2015` preset isn't even the full `ES2015` standard. I think it would make sense to change `preconstruct build` output to use the [the es2015 rollup preset](https://github.com/rollup/rollup/blob/c9bd03d12e96c46122a0372d3bbe9b468cee57da/src/utils/options/options.ts#L79-L85)), as those parts should be `node >= 8.0.0` I think

Preconstruct could also consider enabling `keepDynamicImportAsDynamicImportInCommonJS: true` by default, but that requires `node >= 14.0.0` so it's a harder config change.

Reply:
The Rollup output options aren't really relevant, they're not for down-levelling, they only control a small amount of helper code that Rollup sometimes needs to emit. Updating the target in Babel will make `preconstruct build` output more modern JS

Reply:
Ah. Apologies then, I was completely wrong.